Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s

xush6528 opened the pull request:
Summary:

https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_rpc` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in #29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes #29031

Differential Revision: D18549919

mrshenli commented on test/dist_utils.py:
```diff
@@ -74,22 +73,21 @@ def dist_init(old_test_method=None, setup_rpc=True, clean_shutdown=True):
+                "worker{}".format(rank) for rank in range(self.world_size)
+            }
 
             # Use enough 'num_send_recv_threads' until we fix https://github.com/pytorch/pytorch/issues/26359
```
Remove this line?

xush6528 replied:
Removed.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -33,6 +33,10 @@ PyObject* rpc_init(PyObject* /* unused */) {
 
   auto module = py::handle(rpc_module).cast<py::module>();
 
+  auto rpcAgentOptions =
```
(can be done in a followup PR)

As we are exposing this in API, we should also have docs for it.

xush6528 replied:
Tracked this task in #30032.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -88,6 +92,13 @@ PyObject* rpc_init(PyObject* /* unused */) {
               [&](FutureMessage& fut) { return toPyObj(fut.wait()); },
               py::call_guard<py::gil_scoped_release>());
 
+  py::class_<ProcessGroupRpcAgentOptions>(
```
docs for this one as well (can be done in separate PR)

mrshenli replied:
Any reason this is not using `shared_ptr_class_`?

xush6528 replied:
Didn't do it on purpose. What is the benefit of using `shared_ptr_class_ `?

mrshenli replied:
From [pybind doc](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html):


> If no such holder type template argument is given, the default for a type named Type is std::unique_ptr<Type>, which means that the object is deallocated when Python’s reference count goes to zero.

mrshenli commented on torch/distributed/rendezvous.py:
```diff
@@ -3,6 +3,8 @@
 except ImportError:
     from urlparse import urlparse
 
+import torch
```
if we just need `_six` here, shall we explicitly import _six?

xush6528 replied:
I think so.

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -66,32 +66,49 @@ def _init_rpc_backend(
+        raise RuntimeError("`backend` must be a `backend_registry.BackendType`.")
+
+    if not isinstance(store, dist.Store):
+        raise RuntimeError("`store` must be a c10d::Store. {}".format(store))
```
any reason `c10d::Store` is not quoted?

xush6528 replied:
Should quote it.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -1,24 +1,34 @@
+        backend_name (str): backend string to identify the handler.
+        construct_rpc_agent_options_handler (function):
+            Handler that is invoked when
+            rpc_backen.construct_rpc_agent_options(**dict) is called.
```
backen -> backend

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -1,24 +1,34 @@
-        backend (str): backend string to identify the handler.
-        handler (function): Handler that is invoked when the
+        backend_name (str): backend string to identify the handler.
+        construct_rpc_agent_options_handler (function):
```
this is a pretty long name, but I don't have a better suggestion. Does `rpc_agent_options_factory` work? If we do that `init_backend_handler` will become `backend_factory`.

xush6528 replied:
I don't think it is a backend_factory, it's creating the same type of backend. I can't think of a better name.

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -26,9 +27,8 @@ def init_rpc(
         backend=backend_registry.BackendType.PROCESS_GROUP,
         init_method=None,
```
Shouldn't init_method be subsumed by rpc_agent_options?

xush6528 replied:
@pritamdamania87 It's used by all backbends. So we don't want to put it into `RpcAgentOptions`.

pritamdamania87 commented on torch/distributed/rpc/__init__.py:
```diff
@@ -26,9 +27,8 @@ def init_rpc(
         backend=backend_registry.BackendType.PROCESS_GROUP,
```
General comment, should `self_name` and `self_rank` be `name` and `rank`? For example, `init_process_group` uses `rank` and not `self_rank`.

xush6528 replied:
@mrshenli What do you think? If we all agree. I will do the change.

mrshenli replied:
+1 for this change (drop `self_` in the name)

xush6528 replied:
Updated.

mrshenli reviewed (approved):
LGTM! Thanks!

It looks like CI tests didn't launch properly (only 7 out of ~60 were launched). Please make sure all CI tests pass before landing.

xush6528 commented:
@pytorchbot retest this please

xush6528 committed:
Add `RpcAgentOptions` struct type, which bundles different required arguments for different `RpcAgent`s (#29972)

Summary:
https://github.com/pytorch/pytorch/pull/28226 introduced `worker_to_id` arg to the `def init_rpc` function for other `RpcAgent`. While it's not really used by `ProcessGroupAgent`. Cleanup is wanted for this, as described in https://github.com/pytorch/pytorch/issues/29031.

To adapt to the difference of different `RpcAgent`, adding a `RpcAgentOptions` base classes, which allow leveraging inheritance to add extra fields.

closes https://github.com/pytorch/pytorch/issues/29031
Pull Request resolved: https://github.com/pytorch/pytorch/pull/29972

Differential Revision: D18549919

Pulled By: xush6528

fbshipit-source-id: 7bea84148341d00bbcb29f9bc72707ba5e6ba238
